### ISSUE TITLE
Allow administrators to merge users and visitors

### ISSUE THREAD

**Issue by pgrm** [sandstorm-io/sandstorm#2112]
Hi, currently as a user, you can add new identities to your profile. Which works great. If I logged in by mistake with a wrong email address I've created a *Visitor* user. After logging in with the correct email address, I can link the other email address as an identity to my account and the *Visitor* user disappears. **Awesome!**

Unfortunately, I know of many users for whom it's too complicated. I have for each user / visitor which I invited to my sandstorm instance at least 1 additional account because they mixed up their email addresses, or similar. I'd like to have the possibility as an administrator to do them and me a favor and make our lives easier, by merging those accounts together.

I know this might be a slightly radical feature and it could break a lot of things if used wrongly - maybe and additional authentication token via email could be required before execution or similar. But unfortunately, handling accounts isn't easy, even for people who do study computer science.

**Comment by paulproteus** [comment]
Hi @pgrm ! Let me ping @kentonv to get his attention on this, although it might be a few days or a week until you get a detailed response.

Thanks for filing it!

**Comment by zarvox** [comment]
It's true that managing multiple identities is a bit complicated, and probably more difficult to succeed at than most users can handle, and that it's painfully easy for humans like me to forget what we previously used to log in with, and that we don't really guide users to try to merge accounts if they discover themselves in a state of confusion.  My thoughts are that the problem can be attacked in multiple places, each solving different pieces of the challenge:

* Part of the trouble is in having so many identity providers configured that it becomes easy to forget which one you may have used.  If you want users to have the choice of what they use as their identity provider, then you wind up needing to support them all, but many organizations may only need to use e.g. LDAP login or Google login, and can avoid the complications of multiple identities in the first place.
* Hinting users to log back in as an identity they had used in the past in that browser might help reduce confusion/failure rates here, but only works if you use the same browser and probably the same device.
* Providing UI to help/encourage users to link their accounts with others that share a verified email address might be another automatable, user-driven way to help sort the multiple-accounts problem out.  This probably won't help if you use multiple unrelated email addresses, but it could be a boon for google/github/email mixups.
* Or, as suggested, letting an admin do the action of merging accounts for the user, which is super painless for the user, but has some possibly troublesome trust implications and some user education implications.

While I'm not opposed to the last suggestion on principle, I do think we should be judicious about when we allow/encourage admins to take actions on user accounts that are not their own, especially when trust or privilege boundaries are involved (as they are here).  If a user's account is being modified by an admin, we want to make sure that 1) the user is informed of the change, and the impact, and 2) that there's some form of audit trail.

It's worth noting that even if the admin can resolve the immediate "how do I link my accounts" problem, there's another piece of user education about how to use linked accounts effectively that we might be glossing over -- things like:

* "how do you switch between identities"
* that grains are created and owned by the "current active identity"
* that linking an account means authenticating as either identity gives access to all your grains, etc.)

Just my two cents on the topic. :)

**Comment by pgrm** [comment]
@zarvox you are right, with a larger choice of different providers comes also the possibility to forget about those kind of things. However, only having e-mail is in my experience enough confusion for people to forget if they used their gmail, yahoo, outlook or work address ...

The easiest solution could be, if you have on the account creation page a visible link with some text like "I'm not sure which exact account I used last time - merge with existing account" - ok, the text is terrible, but very descriptive ;)

I've told a guy to go to the sandstorm instance and look at the grain I had shared with him. He knew he had already logged in, but didn't know the email. With that link he could easily fix the problem before creating a duplicate account.

**Comment by zarvox** [comment]
I guess if you have only email token login, then a hint to merge based on matching verified email addresses won't be able to help.

We don't want the server to leak information about other accounts that aren't associated with the logging-in user, so I guess the most we could do is keep a browser-side cache of previously-logged-in identities (under the assumption that if you're reusing the same browser session, you probably have some amount of trust with the person whose browser you're using, if it isn't you).

Then, if you happen to log in as multiple identities from the same browser profile on the same device, we could hint you towards authenticating as each to merge the accounts.

Another thought: if you have the option of telling everyone "just use your work email", you could help people make good choices by putting that sort of information on the [login splash page](https://docs.sandstorm.io/en/latest/administering/faq/#can-i-customize-the-root-page-of-my-sandstorm-install), though admittedly that doesn't help if users are first showing up via sharing links.

I wonder if other services do anything in particular to help people get this right, since presumably the same problem of "I forgot which email address I used with this service" applies to a pretty broad class of services.

**Comment by pgrm** [comment]
Hi, maybe it wasn't clear from my side - of course you shouldn't leak a list of email addresses - just make a button clearly visibly available during the registration process, which shows you that you can merge your accounts. Personally I don't know any user who would have ever looked into their account details. 90% of the users on my server use sandstorm just by using the grains created and shared by the other 10% (yes those numbers are completely made up but it sort of feels like it reflects my reality given the data from /admin/stats). 

So all I'm asking for, is making that feature more prominent in the right moment, and the registration process seems like the right moment for me.

Of course you could add some fuzzy search across the names / email addresses (skipping the domain part) to determine how prominent you want to make that button ("I'm pretty sure you signed up already, don't you want to merge your account?" 😉 ) but that's just icing on the cake, the main problem from my point of view is, that people simply don't know that those features exist and where to find them. 

But maybe it's just my special use case and you have data which prove the opposite in which case I'll just send out an email to all the affected users describing how they should merge their accounts 😃
